File: lc_java/lcJava/stack/RemoveKdigits.java
package lcJava.stack;

import java.util.*;
import java.util.stream.*;

public class RemoveKdigits {
    
    public String removeKdigits(String num, int k) {
        Deque<Character> stack = new ArrayDeque<>();
        
        for (char digit : num.toCharArray()) {
            // 当栈不为空，还有删除次数，且栈顶元素大于当前数字时，删除栈顶
            while (!stack.isEmpty() && k > 0 && stack.peekLast() > digit) {
                stack.removeLast();
                k--;
            }
            stack.addLast(digit);
        }
        
        // 如果还没删完，从后面删
        while (k > 0 && !stack.isEmpty()) {
            stack.removeLast();
            k--;
        }
        
        // 去掉前导零
        StringBuilder sb = new StringBuilder();
        boolean leadingZero = true;
        for (char c : stack) {
            if (c != '0' || !leadingZero) {
                sb.append(c);
                leadingZero = false;
            }
        }
        
        return sb.length() == 0 ? "0" : sb.toString();
    }
    
    // 使用Stream API的函数式版本
    public String removeKdigitsStream(String num, int k) {
        final int[] remainingK = {k};
        Deque<Character> stack = new ArrayDeque<>();
        
        num.chars()
           .mapToObj(c -> (char) c)
           .forEach(digit -> {
               // 移除比当前数字大的栈顶元素
               while (!stack.isEmpty() && remainingK[0] > 0 && stack.peekLast() > digit) {
                   stack.removeLast();
                   remainingK[0]--;
               }
               stack.addLast(digit);
           });
        
        // 如果还有剩余删除次数，从后面删除
        while (remainingK[0] > 0 && !stack.isEmpty()) {
            stack.removeLast();
            remainingK[0]--;
        }
        
        // 去掉前导零并构建结果
        String result = stack.stream()
                            .map(String::valueOf)
                            .collect(Collectors.joining())
                            .replaceFirst("^0+", "");
        
        return result.isEmpty() ? "0" : result;
    }
    
    public static void main(String[] args) {
        RemoveKdigits solution = new RemoveKdigits();
        
        // 测试用例
        String[][] testCases = {
            {"1432219", "3"},  // "1219"
            {"10200", "1"},    // "200"
            {"10", "2"},       // "0"
            {"9", "1"},        // "0"
            {"112", "1"}       // "11"
        };
        
        System.out.println("=== 标准版本 ===");
        for (String[] testCase : testCases) {
            String num = testCase[0];
            int k = Integer.parseInt(testCase[1]);
            String result = solution.removeKdigits(num, k);
            System.out.printf("removeKdigits('%s', %d) = '%s'%n", num, k, result);
        }
        
        System.out.println("\n=== Stream版本 ===");
        for (String[] testCase : testCases) {
            String num = testCase[0];
            int k = Integer.parseInt(testCase[1]);
            String result = solution.removeKdigitsStream(num, k);
            System.out.printf("removeKdigitsStream('%s', %d) = '%s'%n", num, k, result);
        }
    }
}

File: lc_py/lcPy/stack/removeKdigits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        stack = []
        
        for digit in num:
            # 当栈不为空，还有删除次数，且栈顶元素大于当前数字时，删除栈顶
            while stack and k > 0 and stack[-1] > digit:
                stack.pop()
                k -= 1
            stack.append(digit)
        
        # 如果还没删完，从后面删
        while k > 0 and stack:
            stack.pop()
            k -= 1
        
        # 去掉前导零
        result = ''.join(stack).lstrip('0')
        return result if result else '0'

# 函数式版本（使用更多Python特性）
def removeKdigits_functional(num: str, k: int) -> str:
    from functools import reduce
    
    def process_digit(acc, digit):
        stack, remaining_k = acc
        # 移除比当前数字大的栈顶元素
        while stack and remaining_k > 0 and stack[-1] > digit:
            stack.pop()
            remaining_k -= 1
        stack.append(digit)
        return stack, remaining_k
    
    # 使用reduce处理每个数字
    stack, k = reduce(process_digit, num, ([], k))
    
    # 如果还有剩余删除次数，从后面删除
    stack = stack[:len(stack) - k] if k > 0 else stack
    
    # 去掉前导零并返回结果
    result = ''.join(stack).lstrip('0')
    return result or '0'

def main():
    sol = Solution()
    
    # 测试用例
    test_cases = [
        ("12345", 2),
        ("1432219", 3),  # "1219"
        ("10200", 1),    # "200"
        ("10", 2),       # "0"
        ("9", 1),        # "0"
        ("112", 1),      # "11"
    ]
    
    print("=== 标准版本 ===")
    for num, k in test_cases:
        result = sol.removeKdigits(num, k)
        print(f"removeKdigits('{num}', {k}) = '{result}'")
    
    print("\n=== 函数式版本 ===")
    for num, k in test_cases:
        result = removeKdigits_functional(num, k)
        print(f"removeKdigits_functional('{num}', {k}) = '{result}'")

if __name__ == "__main__":
    main()